packages/cli/src/databases/migrations/common/1739549398681-CreateAnalyticsTables.ts

```diff
@@ -0,0 +1,105 @@
+import type { MigrationContext, ReversibleMigration } from '@/databases/types';
+
+const names = {
+	// table names
+	t: {
+		analyticsMetadata: 'analytics_metadata',
+		analyticsRaw: 'analytics_raw',
+		analyticsByPeriod: 'analytics_by_period',
+		workflowEntity: 'workflow_entity',
+		project: 'project',
+	},
+	// column names by table
+	c: {
+		analyticsMetadata: {
+			metaId: 'metaId',
+			projectId: 'projectId',
+			workflowId: 'workflowId',
+		},
+		analyticsRaw: {
+			metaId: 'metaId',
+		},
+		analyticsByPeriod: {
+			metaId: 'metaId',
+			type: 'type',
+			periodUnit: 'periodUnit',
+			periodStart: 'periodStart',
+		},
+		project: {
+			id: 'id',
+		},
+		workflowEntity: {
+			id: 'id',
+		},
+	},
+};
+
+export class CreateAnalyticsTables1739549398681 implements ReversibleMigration {
+	async up({ dbType, schemaBuilder: { createTable, column } }: MigrationContext) {
+		await createTable(names.t.analyticsMetadata)
+			.withColumns(
+				column(names.c.analyticsMetadata.metaId).int.primary.autoGenerate,
+				column(names.c.analyticsMetadata.workflowId).varchar(16),
+				column(names.c.analyticsMetadata.projectId).varchar(36),
+				column('workflowName').varchar(128).notNull,
+				column('projectName').varchar(255).notNull,
+			)
+			.withForeignKey(names.c.analyticsMetadata.workflowId, {
+				tableName: names.t.workflowEntity,
+				columnName: names.c.workflowEntity.id,
+				onDelete: 'SET NULL',
+			})
+			.withForeignKey(names.c.analyticsMetadata.projectId, {
+				tableName: names.t.project,
+				columnName: names.c.project.id,
+				onDelete: 'SET NULL',
+			});
+
+		const timestampDefault =
+			dbType === 'mariadb' || dbType === 'mysqldb' ? 'CURRENT_TIMESTAMP' : "'NOW'";
+
+		await createTable(names.t.analyticsRaw)
+			.withColumns(
+				column('id').int.primary.autoGenerate,
+				column(names.c.analyticsRaw.metaId).int.notNull,
+				column('type').int.notNull,
+				column('value').int.notNull,
+				column('timestamp').timestamp(0).default(timestampDefault).notNull,
+			)
+			.withForeignKey(names.c.analyticsRaw.metaId, {
+				tableName: names.t.analyticsMetadata,
+				columnName: names.c.analyticsMetadata.metaId,
+				onDelete: 'CASCADE',
+			});
+
+		await createTable(names.t.analyticsByPeriod)
+			.withColumns(
+				column('id').int.primary.autoGenerate,
+				column(names.c.analyticsByPeriod.metaId).int.notNull,
+				column(names.c.analyticsByPeriod.type).int.notNull,
+				column('value').int.notNull,
+				column(names.c.analyticsByPeriod.periodUnit).int.notNull,
```

Comment:
Maybe we discussed this already but we could also use `CHAR` and `H`, `D`, `W`. Also could have a column comment and a constraint that restricts that the value can only be one of these

Reply:
Initially it was a `VARCHAR` with `hour`, `week` or `day` and we discussed making it an `INT` or `CHAR`. 

I tried both and `INT` and `CHAR` and CHAR increases the storage needs by 5%:

![image](https://github.com/user-attachments/assets/177a30d1-fd22-479d-b98d-f8ec0e356d7f)

I'd stick to int and map this using getters and setters in the Entity:

```ts
@Entity()
export class AnalyticsByPeriod extends BaseEntity {
	// ...

	@Column({ name: 'periodUnit' })
	periodUnit_: number;

	get periodUnit() {
		if (!isValidType(this.periodUnit_, NumberToPeriodUnit)) {
			throw new ApplicationError('');
		}

		return NumberToPeriodUnit[this.periodUnit_];
	}

	set periodUnit(value: PeriodUnits) {
		this.periodUnit_ = PeriodUnitToNumber[value];
	}
}

```

Reply:
Yeah the different is pretty minimal. In PG there is no difference in size actually:

```sql
CREATE TABLE a (col CHAR NOT NULL);
INSERT INTO a (col)
SELECT SUBSTRING(md5(random()::text) FROM 1 FOR 1)
FROM generate_series(1, 1000000);

CREATE TABLE b (col INT NOT NULL);
INSERT INTO b (col)
SELECT generate_series(1, 1000000);

SELECT pg_total_relation_size('a'), pg_total_relation_size('b');
-- 36290560	36290560
```

Why I thought about using `CHAR` instead is that it would be readable from the value directly if it's hour, day or week if you check the DB. But I'm okay with using INT as well.